extensions/MikeDEV/peerjs.js

```diff
@@ -0,0 +1,929 @@
+            "mikedevpeerjs_whenSpecificPeerDisconnects"
+          );
+        }
+      });
+
+      conn.on("error", (err) => {
+        this.handleChannelError(conn, conn, err);
+      });
+
+      conn.on("data", async (data) => {
+        await this.handleChannelData(conn, conn, JSON.parse(data));
+      });
+    }
+
+    newULID() {
+      return this.ulidGenerator();
+    }
+
+    enableVerboseLogs({ VERBOSE }) {
+      this.verboseLogs = Scratch.Cast.toBoolean(VERBOSE);
+    }
+
+    createPeer({ ID }) {
+      this.peer = new Peer(Scratch.Cast.toString(ID), {
+        config: {
+          // Disable STUN connectivity to mitigate IP leaks
+          iceTransportPolicy: "relay",
+          iceServers: [
+            /* {
+             *  urls: "stun:vpn.mikedev101.cc:3478",
+             * },
+             */
+            {
+              urls: "turn:vpn.mikedev101.cc:3478",
+              username: "free",
+              credential: "free",
+            },
+            /* {
+             *  urls: "stun:vpn.mikedev101.cc:5349",
+             * },
+             */
+            {
+              urls: "turn:vpn.mikedev101.cc:5349",
+              username: "free",
+              credential: "free",
+            },
+          ],
+        },
+        // Only enable verbose logs if the user wants it - This can get very laggy if left enabled by accident
+        debug: this.verboseLogs ? 3 : 2,
+      });
+
+      this.peer.errorInfo = "";
+
+      this.peer.on("open", (id) => {
+        if (id === Scratch.Cast.toString(ID)) {
+          Scratch.vm.runtime.startHats("mikedevpeerjs_whenPeerCreated");
+        }
+      });
+
+      this.peer.on("connection", (conn) => {
+        conn.idCounter = 2;
+        conn.channels = new Map();
+        this.dataConnections.set(conn.peer, conn);
+        this.handleDataConnection(conn);
+      });
+
+      this.peer.on("call", async (call) => {
+        if (!this.hasMicPerms || this.voiceConnections.has(call.peer)) {
+          call.close();
+          return;
+        }
+        this.ringingPeers.set(call.peer, call);
+        Scratch.vm.runtime.startHats("mikedevpeerjs_whenPeerRings");
+      });
+
+      this.peer.on("close", () => {
+        Scratch.vm.runtime.startHats("mikedevpeerjs_whenPeerDestroyed");
+      });
+
+      this.peer.on("disconnected", () => {
+        Scratch.vm.runtime.startHats("mikedevpeerjs_whenPeerDisconnected");
+      });
+
+      this.peer.on("error", (err) => {
+        console.log("Peer error: " + err);
+        this.peer.errorInfo = err;
+        Scratch.vm.runtime.startHats("mikedevpeerjs_whenPeerHasError");
+      });
+    }
+
+    readPeerErrorInfo() {
+      if (!this.peer) return "";
+      return this.peer.errorInfo;
+    }
+
+    connectToPeer({ ID }) {
+      if (!this.isPeerConnected()) return;
+      if (this.dataConnections.has(Scratch.Cast.toString(ID))) return;
+      const conn = this.peer.connect(Scratch.Cast.toString(ID), {
```

Comment:
Instead of calling `Scratch.Cast.toString(ID)` multiple times, just make a quick `const` of it so you have a static one you can fetch for efficiency, or do `ID = Scratch.Cast.toString(ID)`

Reply:
Makes sense.